iroh-net/src/hp/netcheck/reportgen.rs

```diff
@@ -0,0 +1,1006 @@
+    last_report: Option<Arc<Report>>,
+    /// The portmapper client, if there is one.
+    port_mapper: Option<portmapper::Client>,
+    skip_external_network: bool,
+    /// The DERP configuration.
+    derp_map: DerpMap,
+    /// Socket to send IPv4 STUN requests from.
+    stun_sock4: Option<Arc<UdpSocket>>,
+    /// Socket so send IPv6 STUN requests from.
+    stun_sock6: Option<Arc<UdpSocket>>,
+
+    // Internal state.
+    /// Whether we're doing an incremental report.
+    incremental: bool,
+    /// The report being built.
+    report: Report,
+    /// The hairping actor.
+    hairpin_actor: hairpin::Client,
+    /// Which tasks the reportgen actor is still waiting on.
+    ///
+    /// This is essentially the summary of all the work the reportgen actor is doing.
+    outstanding_tasks: OutstandingTasks,
+}
+
+impl Actor {
+    fn addr(&self) -> Addr {
+        Addr {
+            sender: self.msg_tx.clone(),
+        }
+    }
+
+    #[instrument(name = "reportgen.actor", skip_all)]
+    async fn run(&mut self) {
+        match self.run_inner().await {
+            Ok(_) => debug!("reportgen actor finished"),
+            Err(err) => {
+                error!("reportgen actor failed: {err:#}");
+                self.netcheck
+                    .send(netcheck::Message::ReportAborted)
+                    .await
+                    .ok();
+            }
+        }
+    }
+
+    /// Runs the main reportgen actor logic.
+    ///
+    /// This actor runs by:
+    ///
+    /// - Creates a hairpin actor.
+    /// - Creates a captive portal future.
+    /// - Creates ProbeSet futures in a group of futures.
+    /// TODO: pinger
+    /// - Runs a main loop:
+    ///   - Drives all the above futures.
+    ///   - Receives actor messages (sent by those futures).
+    ///   - Updates the report, cancels unneeded futures.
+    /// - Sends the report to the netcheck actor.
+    async fn run_inner(&mut self) -> Result<()> {
+        debug!(
+            port_mapper = %self.port_mapper.is_some(),
+            skip_external_network=%self.skip_external_network,
+            "reportstate actor starting",
+        );
+
+        let start_time = Instant::now();
+        self.report.os_has_ipv6 = super::os_has_ipv6().await;
+
+        let mut port_mapping = self.prepare_portmapper_task();
+        let mut captive_task = self.prepare_captive_portal_task();
+        let mut probes = self.prepare_probes_task().await?;
+
+        loop {
+            trace!(awaiting = ?self.outstanding_tasks, "tick; awaiting tasks");
+            if self.outstanding_tasks.all_done() {
+                debug!("all tasks done");
+                break;
+            }
+            let remaining_time = OVERALL_PROBE_TIMEOUT.saturating_sub(start_time.elapsed());
+            let remaining_probe_time = STUN_PROBE_TIMEOUT.saturating_sub(start_time.elapsed());
+
+            tokio::select! {
+                _ = tokio::time::sleep(remaining_time) => {
```

Comment:
This seems inefficient (I might be missing sth), you create a new `Sleep` on every loop, but you could just do this

```rust
let total_timer = tokio::time::sleep(OVERALL_PROBE_TIMEOUT);
tokio::pin!(total_timer);

let stun_timer =  tokio::time::sleep(STUN_PROBE_TIMEOUT);
tokio::pin!(total_timer);

loop {
  tokio::select! {
    _ = &mut total_timer => {
    }
    _ = &mut stun_timer => {
    }
  }
}
```

Reply:
no, it was me who was missing something :)